analysis/pseudobulk-bulk-prediction/notebooks/perform-ora.Rmd

```diff
@@ -0,0 +1,270 @@
+---
+title: "Over-representation analysis with consensus cell types"
+author: "Stephanie J. Spielman"
+date: "`r Sys.Date()`"
+output:
+  html_notebook:
+    toc: true
+    toc_depth: 3
+    code_folding: hide
+params:
+  project_id: "SCPCP000001"
+  summary_function: "median"
+  sd_threshold: 2.5
+  reps: 1000
+---
+
+## Setup
+
+
+```{r setup}
+renv::load()
+
+library(ggplot2)
+theme_set(theme_bw())
+
+set.seed(2025)
+```
+
+### Paths
+
+```{r paths}
+result_dir <- here::here("analysis", "pseudobulk-bulk-prediction", "results")
+model_dir <- here::here(result_dir, "models")
+data_dir <- here::here("analysis", "pseudobulk-bulk-prediction", "data")
+
+model_data_file <- file.path(
+  model_dir, 
+  glue::glue("{params$project_id}_bulk-pseudobulk-model_threshold-0.rds")
+) 
+geneset_file <- file.path(
+  data_dir, 
+  glue::glue("{params$project_id}_panglao-genesets.tsv")
+) 
+
+output_file <- file.path(
+  result_dir, 
+  glue::glue("{params$project_id}_ORA-odds-ratios.tsv")
+)
+```
+
+### Functions
+
+```{r}
+# Function to calculate our odds ratio of interest from a data frame with expected columns:
+# in_geneset (logical), outlier (logical), n (integer)
+calculate_or <- function(df) {
+  # (in geneset & outlier) / (in geneset & not outlier) 
+  # divided by
+  # (not in geneset & outlier) / (not in geneset & not outlier) 
+  # This is:
+  # The odds a cell type marker gene is an outlier compared to non-cell-type marker genes
+  odds_num <- (dplyr::filter(df, in_geneset, outlier) |> dplyr::pull(n)) / (dplyr::filter(df, in_geneset, !outlier) |> dplyr::pull(n))
+  odds_denom <- (dplyr::filter(df, !in_geneset, outlier) |> dplyr::pull(n)) / (dplyr::filter(df, !in_geneset, !outlier) |> dplyr::pull(n))
+  
+  return(odds_num / odds_denom)
+}
```

Comment:
As discussed in our meeting earlier this week, ORs take everyone a second 😄 

As discussed in Slack, this is _equivalent_ to how I think about the problem (odds an outlier is in a gene set compared to non-outliers), but your way is also how @jashapiro thinks about the problem.